presto-parquet/src/main/java/io/prestosql/parquet/reader/PrimitiveColumnReader.java

```diff
@@ -78,32 +92,80 @@ protected boolean isValueNull()
+                    return new LongColumnReader(descriptor);
                 }
-                return createDecimalColumnReader(descriptor).orElse(new LongColumnReader(descriptor));
-            case INT96:
-                return new TimestampColumnReader(descriptor);
-            case FLOAT:
-                return new FloatColumnReader(descriptor);
-            case DOUBLE:
-                return new DoubleColumnReader(descriptor);
-            case BINARY:
-                return createDecimalColumnReader(descriptor).orElse(new BinaryColumnReader(descriptor));
-            case FIXED_LEN_BYTE_ARRAY:
-                return createDecimalColumnReader(descriptor)
-                        .orElseThrow(() -> new PrestoException(NOT_SUPPORTED, " type FIXED_LEN_BYTE_ARRAY supported as DECIMAL; got " + descriptor.getPrimitiveType().getOriginalType()));
-            default:
-                throw new PrestoException(NOT_SUPPORTED, "Unsupported parquet type: " + descriptor.getPrimitiveType().getPrimitiveTypeName());
+            }
+        }
+
+        if (prestoType instanceof DecimalType) {
+            if (parquetDecimalType.isPresent() && ((DecimalType) prestoType).getScale() == parquetDecimalType.get().getScale()) {
+                // TODO allow rescaling
+                return decimalColumnReader.orElseThrow(() -> new VerifyException("decimalColumnReader not present when parquetDecimalType is present"));
+            }
+        }
+
+        if (prestoType == RealType.REAL && descriptor.getPrimitiveType().getPrimitiveTypeName() == PrimitiveTypeName.FLOAT) {
+            return new FloatColumnReader(descriptor);
+        }
+
+        if (prestoType == DoubleType.DOUBLE && descriptor.getPrimitiveType().getPrimitiveTypeName() == PrimitiveTypeName.DOUBLE) {
+            return new DoubleColumnReader(descriptor);
+        }
+
+        if (prestoType == DateType.DATE && descriptor.getPrimitiveType().getPrimitiveTypeName() == PrimitiveTypeName.INT32) {
+            // TODO verify no truncation occurs
+            return new IntColumnReader(descriptor);
+        }
+
+        if (prestoType == TimestampType.TIMESTAMP && descriptor.getPrimitiveType().getPrimitiveTypeName() == PrimitiveTypeName.INT96) {
+            return new TimestampColumnReader(descriptor);
+        }
+
+        if (prestoType == TimestampType.TIMESTAMP && descriptor.getPrimitiveType().getPrimitiveTypeName() == PrimitiveTypeName.INT64 && descriptor.getPrimitiveType().getOriginalType() == OriginalType.TIMESTAMP_MICROS) {
+            return new TimestampMicrosColumnReader(descriptor);
         }
+
+        if (prestoType instanceof VarcharType && descriptor.getPrimitiveType().getPrimitiveTypeName() == PrimitiveTypeName.BINARY) {
+            return new BinaryColumnReader(descriptor);
+        }
+
+        if (prestoType instanceof CharType && descriptor.getPrimitiveType().getPrimitiveTypeName() == PrimitiveTypeName.BINARY) {
+            return new BinaryColumnReader(descriptor);
```

Comment:
do we make sure that trailing spaces are truncated? Can we assume they are (probably not)?

Reply:
> do we make sure that trailing spaces are truncated? Can we assume they are (probably not)?

they are